poetry/core/masonry/builders/builder.py

```diff
@@ -268,3 +302,38 @@ def temporary_directory(cls, *args, **kwargs):
             yield name
 
             shutil.rmtree(name)
+
+
+class BuildIncludeFile:
+    def __init__(
+        self,
+        path,  # type: Path
+        source_root=None,  # type: Optional[Path]
+    ):
+        """
+        :param path: a full path to the file to be included
+        :param source_root: the root path to resolve to
+        """
+        self.path = Path(path).resolve()
+        self.source_root = None if not source_root else Path(source_root).resolve()
+        if not self.path.is_absolute() and self.source_root:
+            self.path = (self.source_root / self.path).resolve()
+
+    def __eq__(self, other):  # type: (Union[BuildIncludeFile, Path]) -> bool
+        if hasattr(other, "path"):
+            return self.path == other.path
+        return self.path == other
+
+    def __ne__(self, other):  # type: (Union[BuildIncludeFile, Path]) -> bool
```

Comment:
Does this need to be explicitly defined?

Reply:
Yes, otherwise the hashing function for the set does not work properly: there can be cases where a `Path` is checked if it is in a set of `BuildIncludeFile` objects.